linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,350 @@
+
+`threadpool_destroy` shall implicit close if `threadpool` is open. 
+
+`threadpool_destroy` shall free the resources associated with the threadpool handle. 
+
+`threadpool_destroy` shall destroy the semphore by calling `sem_destroy`.
+
+`threadpool_destroy` shall destroy the SRW lock by calling `srw_lock_destroy`.
+
+### threadpool_open_async
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_open_async, THREADPOOL_HANDLE, threadpool, ON_THREADPOOL_OPEN_COMPLETE, on_open_complete, void*, on_open_complete_context);
+```
+
+`threadpool_open_async` opens the threadpool asynchronously.
+
+If `threadpool` is `NULL`, `threadpool_open_async` shall fail and return a non-zero value. 
+
+`threadpool_open_async` shall call `sm_open_begin`. 
+
+`threadpool_open_async` shall create the threads for `threadpool` using `ThreadAPI_Create`.
+
+If `threadpool` has already been opened, `threadpool_open_async` shall fail and return a non-zero value.
+
+If one of the thread creation fails, `threadpool_open_async` shall fail and return a non-zero value, terminate all threads already created.
+
+Otherwise, `threadpool_open_async` shall shall call `sm_open_end` with true for success.
+
+`threadpool_open_async` shall succeed and return zero, indicate open success to the user by calling the `on_open_complete` callback with `THREADPOOL_OPEN_OK`.
+
+### threadpool_close
+
+```C
+MOCKABLE_FUNCTION(, void, threadpool_close, THREADPOOL_HANDLE, threadpool);
+```
+
+`threadpool_close` closes the threadpool.
+
+If `threadpool` is `NULL`, `threadpool_close` shall fail and return.
+
+Otherwise, `threadpool_close` shall join all threads in the `threadpool`.
+
+`threadpool_close` shall call `sm_close_begin`.
+
+`threadpool_close` shall call `sm_close_end`.
+
+### threadpool_schedule_work
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work, THREADPOOL_HANDLE, threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
+```
+
+`threadpool_schedule_work` schedule the threadpool to work.
+
+If `threadpool` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+If `work_function` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall get the threadpool state by calling `interlocked_add`.
+
+`threadpool_schedule_work` shall call `sm_exec_begin`.
+
+If `sm_exec_begin` returns `SM_EXEC_REFUSED`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall acquire the SRW lock in shared mode by calling `srw_lock_acquire_shared`.
+
+`threadpool_schedule_work` shall increment the `insert_pos`.
+
+If task state is `TASK_NOT_USED`, `threadpool_schedule_work` shall set the current task state to `TASK_INITIALIZING`.
+
+Otherwise, `threadpool_schedule_work` shall release the shared SRW lock by calling `srw_lock_release_shared` and increase `task_array` capacity by calling `reallocate_threadpool_array`.
```

Comment:
>reallocate_threadpool_array [](http://example.com/codeflow?start=156&length=27)

This is an internal implementation detail that should not be at all here in the specs. Please remove it.

Reply:
deleted